replace all_gather with more efficient collective api _all_gather_base

zhaojuanmao opened the pull request:
Summary: _all_gather_base saved copies in all_gather, so it is more efficient

Test Plan: unit test

Reviewed By: SciPioneer

Differential Revision: D28227193

albanD commented:
@mrshenli do you know who would be best to review this?

SciPioneer commented:
Please address the test failures.

ngimel commented on torch/nn/modules/_functions.py:
```diff
@@ -27,15 +27,31 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
+        if process_group._get_backend_name() == 'nccl':
+            # world_size * (2C + 1)
+            combined_size = combined.numel()
+            combined_flat = torch.zeros(1,
```
The values will be overwritten anyway, so you can use torch.empty here?

zhaojuanmao committed:
replace all_gather with more efficient collective api _all_gather_base (#57769)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/57769

_all_gather_base saved copies in all_gather, so it is more efficient

Test Plan: unit test

Reviewed By: SciPioneer

Differential Revision: D28227193

fbshipit-source-id: 68b837f4aedc8764478f67423533bb41dac73189
